### PR TITLE
D3log realization sinks api

### PR DESCRIPTION
**Pull request by krs85**
This pull request continues the refactoring of the Realization interface by extending functionality to the sink side. These additions include the ability to add and remove both file and tcp sender sinks. The functionality also includes the ability to add and remove sink accumulators. The following is a list of functionality added:

- Implements add_sink() and remove_sink() for both file sinks and tcp sender sinks.
- Implements add_sink_accumulator() and remove_sink_accumulator().
- Refactors add_tcp_senders() and add_file_sinks() to use this new API.
- Fixes existing bug where a new stream was being added to the server for every sink, instead of a new stream added to the server for each accumulator.
- Adds multiple functions to Realization for the purpose of unit testing. Unit testing for Realization is done in server_api, and because elements of Realization are private, a few helper functions were needed.
- Adds unit tests to exercise the new changes to the Realization API.

### CODE REVIEW

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 1]
```diff
@@ -405,13 +301,218 @@ where
...
+                        let tcp_sender = TcpSender::new(*address)
+                            .map_err(|e| format!("failed to create TcpSender socket: {}", e))?;
+                        let sink = Arc::new(Mutex::new(tcp_sender));
+
+                        // Subscribe the accumulator to this sink.
+                        let subscription = accumulator
+                            .lock()
+                            .unwrap()
+                            .subscribe(Box::new(sink.clone()))
+                            .map_err(|_| "failed to subscribe file sink to accumulator")?;
```
Comment (ryzhyk) [inline rust/template/distributed_datalog/src/instantiate.rs:365]: file sink -> TCP sink

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 2]
```diff
@@ -405,13 +301,218 @@ where
...
+            if sink_map.is_empty() {
+                // Disconnect accumulator from server.
+                // First, unsubscribe the stream from the accumulator.
+                let _ = stream.unsubscribe(&());
+                // Next, remove the stream from the server.
+                server.remove_stream(stream);
+
+                let mut accum = accumulator.lock().unwrap();
+                // Clear the accumulator.
+                if accum.clear().is_ok() {
```
Comment (ryzhyk) [inline rust/template/distributed_datalog/src/instantiate.rs:401]: `accum` is the last reference to this accumulator, right? So shouldn't it clear itself internal state as it goes out of scope?

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 3]
```diff
@@ -405,13 +301,218 @@ where
...
+                    Ok(())
+                } else {
+                    Err("Cannot remove accumulator, failed to clear".to_string())
+                }
+            } else {
+                Err("Cannot remove accumulator, transaction in progress".to_string())
```
Comment (ryzhyk) [inline rust/template/distributed_datalog/src/instantiate.rs:407]: Why is this a transaction-in-progress error as opposed to just accumulator still having sinks attached to it?

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 4]
```diff
@@ -405,13 +301,218 @@ where
...
+    /// Used for testing.
+    pub fn txn_subscription_exists(&self, id: usize) -> bool {
+        self._txnmux.subscription_exists(id)
+    }
+
+    /// Checks that the given sink both has an associated accumulator and
+    /// that said accumulator is subscribed to the sink.
+    /// Used for testing.
+    pub fn accumulator_is_subscribed_to_sink(&self, rel_ids: BTreeSet<RelId>, sink: &Sink) -> bool {
+        let (_, _, sink_map) = self._sinks.get(&rel_ids).unwrap();
```
Comment (ryzhyk) [inline rust/template/distributed_datalog/src/instantiate.rs:473]: Please annotate all methods used for testing with `#[cfg(test)]`.

File: test/datalog_tests/server_api/tests/realization_tests.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        assert_eq!(
+            realization.add_sink(&sink, rel_ids.clone(), &mut server),
+            Ok(())
+        );
+        // Check that removing the sink accumulator returns an error.
+        assert!(realization
+            .remove_sink_accumulator(rel_ids, &mut server)
+            .is_err());
+    }
+}
```
Comment (ryzhyk) [inline test/datalog_tests/server_api/tests/realization_tests.rs:154]: These tests look good, but we should also test the dataflow through the system, e.g., that connecting a new sink to an accumulatator brings it up to a correct state.

Let's wrap up this PR and try to add more tests as a separate PR.

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 6]
```diff
@@ -316,20 +196,36 @@ where
...
-    _sinks: HashMap<BTreeSet<RelId>, Vec<(SinkRealization<P::Convert>, usize)>>,
 }
 
 impl<P> Realization<P>
 where
     P: Send + DDlog + 'static,
     P::Convert: Send + DDlogConvert,
 {
+    /// Instantiates a new, default Realization.
+    #[allow(clippy::new_without_default)]
```
[the remark this answers is not in the document]
Comment (krs85) [reply rust/template/distributed_datalog/src/instantiate.rs:216]: I think this was a mental TODO that I forgot about... but you're totally right, will fix!

File: rust/template/distributed_datalog/src/instantiate.rs [hunk 7]
```diff
@@ -405,13 +301,218 @@ where
...
+    /// Return an error if unable to find a sink map containing the sink.
+    pub fn remove_sink(&mut self, sink: &Sink) -> Result<(), String> {
+        for (accumulator, _, sink_map) in self._sinks.values_mut() {
+            if sink_map.contains_key(sink) {
+                let (_, subscription) = sink_map.remove(sink).unwrap();
+                let _ = accumulator.unsubscribe(&subscription);
+                return Ok(());
+            }
+        }
+        Err("failed to remove sink from realization".to_string())
```
[the remark this answers is not in the document]
Comment (krs85) [reply rust/template/distributed_datalog/src/instantiate.rs:317]: Makes sense, will fix.

File: rust/template/distributed_datalog/src/instantiate.rs (hunk 1 above)
Comment (krs85) [reply rust/template/distributed_datalog/src/instantiate.rs:365]: Makes sense, will fix.

File: test/datalog_tests/server_api/tests/realization_tests.rs (hunk 5 above)
Comment (krs85) [reply test/datalog_tests/server_api/tests/realization_tests.rs:154]: Sounds good to me.

**Pull request merged**